src/lightning/pytorch/strategies/xla_fsdp.py

```diff
@@ -0,0 +1,478 @@
+            self._save_checkpoint_shard(path, converted_state, storage_options)
+
+        if self._state_dict_type == "full":
+            ckpt_prefix = str(path / "checkpoint")
+            ckpt_suffix = "_rank-*-of-*.pth"
+            if len(parallel_devices) != self.world_size:  # multihost
+                raise OSError(
+                    "Multihost setups do not have a shared filesystem, so the checkpoint shards cannot be consolidated"
+                    " into a single checkpoint after saving them. Please switch to"
+                    " `XLAFSDPStrategy(state_dict_type='sharded')`. TIP: You can consolidate them manually by getting"
+                    " them together into a single directory and running `python -m"
+                    f" torch_xla.distributed.fsdp.consolidate_sharded_ckpts --ckpt_prefix {ckpt_prefix!r} --ckpt_suffix"
+                    f" {ckpt_suffix!r} --save_path 'path/to/consolidated.ckpt'`."
+                )
+
+            from torch_xla.distributed.fsdp import consolidate_sharded_model_checkpoints
+
+            self.barrier("before_ckpt_consolidation")
+            if self.is_global_zero:
+                save_path = path.parent / "consolidated.ckpt"
+                # save consolidated checkpoint separate to the shards
+                consolidate_sharded_model_checkpoints(ckpt_prefix, ckpt_suffix, str(save_path))
+                # remove the shards directory
+                self.checkpoint_io.remove_checkpoint(path)
+            self.barrier("after_ckpt_consolidation")
+
+    def _save_checkpoint_shard(
+        self,
+        path: Path,
+        converted_state: Dict[str, Union[Module, Optimizer, Any]],
+        storage_options: Optional[Any],
+    ) -> None:
+        self.checkpoint_io.save_checkpoint(
+            converted_state,
+            path / f"checkpoint_rank-{self.global_rank:08d}-of-{self.world_size:08d}.pth",
+            storage_options=storage_options,
+        )
+
+    def load_checkpoint(
+        self,
+        path: _PATH,
+        strict: bool = True,
+    ) -> Dict[str, Any]:
+        """Given a folder, load the contents from a checkpoint and restore the state of the given objects.
+
+        The strategy currently only supports saving and loading sharded checkpoints which are stored in form of a
+        directory of multiple files rather than a single file.
+
+        """
+        raise NotImplementedError("`load_checkpoint` is not implemented for `XLAFSDPStrategy.")
+
+    @classmethod
+    def register_strategies(cls, strategy_registry: _StrategyRegistry) -> None:
+        strategy_registry.register("xla_fsdp", cls, description=cls.__name__)
+
+    def lightning_module_state_dict(self) -> Dict[str, Any]:
+        assert self.model is not None
+        assert self.optimizers is not None
+        return {
+            "model": self.model.state_dict(),
+            "shard_metadata": self.model.get_shard_metadata(),
+            "optimizer": self.optimizers[0].state_dict(),
```

Comment:
Why did you also include the optimizer state here? This wouldn't be available during non-fit. The strategy has an `optimizer_state` for this.

Reply:
I added `optimizer` here since when running `trainer.fit()`, the checkpointing is called from https://github.com/Lightning-AI/lightning/blob/60eecd15ba5b5b9ca40a6f22d302b78708e8b4d8/src/lightning/pytorch/strategies/launchers/xla.py#L152-L156.

If we don't include the optimizer where you highlighted it, then it won't get saved during `trainer.fit()`

Reply:
I don't get it. It will still get saved for `fit` in https://github.com/Lightning-AI/lightning/blob/master/src/lightning/pytorch/trainer/connectors/checkpoint_connector.py#L442-L448

Reply:
From an offline conversation, (2) is used by `XLALauncher` to update the model weights after the spawned subprocesses in `xmp.spawn()` are finished. However, as shown in [https://github.com/Lightning-AI/lightning/blob/984f49f7195ddc67e961c7c498ee6e19fc0[…]5/src/lightning/pytorch/strategies/launchers/multiprocessing.py](https://github.com/Lightning-AI/lightning/blob/984f49f7195ddc67e961c7c498ee6e19fc0cecb5/src/lightning/pytorch/strategies/launchers/multiprocessing.py#L167), this only loads the model weights, so the optimizer is not needed.